File: code/12-JavaCodeLoop/JavaTDDLoop/src/main/java/com/example/domain/SecretNumberGenerator.java
package com.example.domain;

import com.example.valueobjects.Interval;

import java.util.random.RandomGenerator;

/**
 * Generates a random secret number within a given {@link Interval}.
 */
public class SecretNumberGenerator {

    private final RandomGenerator random;

    public SecretNumberGenerator() {
        this(RandomGenerator.getDefault());
    }

    protected SecretNumberGenerator(RandomGenerator random) {
        this.random = random;
    }

    public int generate(Interval interval) {
        int bound = interval.upper() - interval.lower() + 1;
        return interval.lower() + random.nextInt(bound);
    }

    public static SecretNumberGenerator create() {
        return new SecretNumberGenerator();
    }
}

File: code/13-Java2Python/python/domain/secret_number_generator.py
import random

from valueobjects.interval import Interval


class SecretNumberGenerator:
    """Generates a random secret number within a given Interval."""

    def __init__(self, rng: random.Random | None = None) -> None:
        self._rng = rng if rng is not None else random.Random()

    def generate(self, interval: Interval) -> int:
        bound = interval.upper - interval.lower + 1
        return interval.lower + self._rng.randrange(bound)

    @staticmethod
    def create() -> "SecretNumberGenerator":
        return SecretNumberGenerator()
